### PR TITLE
ch-grow: add subcommands and merge ch-tug

### PR DESCRIPTION
**Pull request by reidpr**
This PR introduces subcommands for `ch-grow`, and it removes the `ch-tug` script and moves that functionality to `ch-grow pull`. What used to be plain `ch-grow` is now `ch-grow build`, and I moved most of that code to `build.py`.

Along the way, it also addresses #839 and #841.

The key motivation is clarity of communication and maintainability. Examples:

* We have a lot of bugs labeled `ch-grow/ch-tug`.
* It's hard to communicate that `ch-tug` has a special relationship with `ch-grow` that the other scripts do not.
* Why is `ch-tug` a separate script but `ch-grow --list` is not?
* `ch-tug` has a bunch of duplicate code and the command line options have already diverged from `ch-grow`.

All these problems would only increase as the unprivileged build functionality grows, e.g. #740 which proposes a new script `ch-nudge` for image pushing.

### CODE REVIEW

File: doc/ch-grow_desc.rst [hunk 1]
```diff
@@ -3,89 +3,146 @@ Synopsis
 
 ::
 
-   $ ch-grow [OPTIONS] [-t TAG] [-f DOCKERFILE] CONTEXT
+   $ ch-grow [...] build [-t TAG] [-f DOCKERFILE] [...] CONTEXT
+   $ ch-grow [...] print-storage
+   $ ch-grow [...] pull [...] IMAGE_REF [IMAGE_DIR]
+   $ ch-grow { --help | --version | --dependencies }
```
Comment (j-ogas) [inline doc/ch-grow_desc.rst:9]: I think this is a step in the right direction. However, I think we would eventually want to abstract yet another layer. See below.

```
   $ ch [...] build [-b|--builder] [-t TAG] [-f DOCKERFILE] [...] CONTEXT
   $ ch [...] list [...] # e.g., print-storage, images, etc.
   $ ch [...] pull [...] IMAGE_REF [IMAGE_DIR]
   $ ch [...] push [...] IMAGE_REF DEST_IMG_REF
   $ ch [...] run [OPTIONS] IMAGE -- CMDS
   $ ch { --help | --version | --dependencies }
```

We have a lot of manual tools right now. Condensing some of them will dramatically change the workflow. How annoying would it be to have the interface changed (admittedly for the better) just to be changed again down the road?

Comment (reidpr) [reply doc/ch-grow_desc.rst:9]: This is a good point, though I'm still not convinced we want to do subcommands across all of Charliecloud. In particular I see running vs. building/pushing/pulling/etc. as separate operations.

As for doing it right now, I'm seeing maintainability problems already. E.g. `ch-tug` has a bunch of duplicate code and the command line options have diverged from `ch-grow`, and we have a bunch of bug reports titled "ch-grow/ch-tug: blah blah blah"; the latter doesn't scale to `ch-nudge` at all. Why does `ch-grow --list` exist but `ch-nudge` is a separate script? Etc.

**Review by j-ogas — APPROVED** [review]

File: bin/ch-grow.py.in [hunk 2]
```diff
@@ -1,722 +1,163 @@
...
 
+   # list
+   help="list images in storage"
+   sp = sps.add_parser("list", help=help, description=help,
+                       formatter_class=ch.HelpFormatter)
+   sp.set_defaults(func=misc.list_)
+   add_opts(sp, common_opts, True)
+
+   # pull
+   help="pull image from remote repository to local fileysystem"
```
Comment (j-ogas) [inline bin/ch-grow.py.in:134]:
```suggestion
   help="pull image from remote repository to local filesystem"
```

File: doc/ch-grow_desc.rst [hunk 3]
```diff
@@ -25,67 +59,96 @@ location is configurable. In descending order of priority:
...
-.. note::
+The storage directory can reside on any filesystem. However, it contains lots
+of small files and metadata traffic can be intense. For example, the
+Charliecloud test suite uses approximately 400,000 files and directories in
+the storage directory as of this writing. Place it on a filesystem appropriate
+for this; tmpfs'es such as :code:`/var/tmp` are a good choice if you have
+enough RAM (:code:`/tmp` is not recommended because :code:`ch-run` bind-mounts
+it into containers by default). Network filesystems, especially Lustre, are
+typically bad choices. This is a site-specific question and your local support
+will likely have strong opinions.
```
[the remark this answers is not in the document]
Comment (rstyd) [reply doc/ch-grow_desc.rst:70]: I agree this could cause some issues for someone unaware.

**Review by rstyd — APPROVED** [review]
Looks good to me

**Pull request merged**